Userland/Libraries/LibJS/AST.cpp

```diff
@@ -444,129 +527,315 @@ Value ForStatement::execute(Interpreter& interpreter, GlobalObject& global_objec
-{
-    if (is<VariableDeclaration>(node)) {
-        auto& variable_declaration = static_cast<VariableDeclaration const&>(node);
-        VERIFY(!variable_declaration.declarations().is_empty());
-        if (variable_declaration.declaration_kind() != DeclarationKind::Var) {
-            wrapper = create_ast_node<BlockStatement>(node.source_range());
-            interpreter.enter_scope(*wrapper, ScopeType::Block, global_object);
+        // i. If destructuring is false, then
+        if (!destructuring) {
+            VERIFY(lhs_reference.has_value());
+            if (lhs_kind == LexicalBinding)
+                lhs_reference->initialize_referenced_binding(global_object, next_value);
+            else
+                lhs_reference->put_value(global_object, next_value);
+            if (auto* exception = interpreter.exception())
+                return throw_completion(exception->value());
+            return {};
+        }
+
+        // j. Else,
+        if (lhs_kind == Assignment) {
+            VERIFY(pattern_lhs);
+            return interpreter.vm().destructuring_assignment_evaluation(*pattern_lhs, next_value, global_object);
         }
-        variable_declaration.execute(interpreter, global_object);
-        return variable_declaration.declarations().first().target();
+        VERIFY(expression_lhs && is<VariableDeclaration>(*expression_lhs));
+        auto& for_declaration = static_cast<VariableDeclaration const&>(*expression_lhs);
+        auto& binding_pattern = for_declaration.declarations().first().target().get<NonnullRefPtr<BindingPattern>>();
+        VERIFY(lhs_kind == VarBinding || iteration_environment);
+
+        // At this point iteration_environment is undefined if lhs_kind == VarBinding which means this does both
+        // branch j.ii and j.iii because ForBindingInitialization is just a forwarding call to BindingInitialization.
+        return interpreter.vm().binding_initialization(binding_pattern, next_value, iteration_environment, global_object);
     }
+};
+
+// 14.7.5.5 Runtime Semantics: ForInOfLoopEvaluation, https://tc39.es/ecma262/#sec-runtime-semantics-forinofloopevaluation
+// 14.7.5.6 ForIn/OfHeadEvaluation ( uninitializedBoundNames, expr, iterationKind ), https://tc39.es/ecma262/#sec-runtime-semantics-forinofheadevaluation
+// This method combines ForInOfLoopEvaluation and ForIn/OfHeadEvaluation for similar reason as ForIn/OfBodyEvaluation, to prevent code duplication.
+// For the same reason we also skip step 6 and 7 of ForIn/OfHeadEvaluation as this is done by the appropriate for loop type.
+static ThrowCompletionOr<ForInOfHeadState> for_in_of_head_execute(Interpreter& interpreter, GlobalObject& global_object, Variant<NonnullRefPtr<ASTNode>, NonnullRefPtr<BindingPattern>> lhs, Expression const& rhs)
+{
+
```

Comment:
Please don't add a newline after the function body start (or any scope start for that matter)